messaging/src/main/java/org/axonframework/common/infra/JacksonComponentDescriptor.java

```diff
@@ -0,0 +1,203 @@
+
+import java.util.Collection;
+import java.util.IdentityHashMap;
+import java.util.Map;
+
+/**
+ * A {@link ComponentDescriptor} implementation that uses Jackson's {@link ObjectMapper} to create JSON representations
+ * of components. This implementation produces a clean, hierarchical JSON structure.
+ * <p>
+ * This implementation supports circular references between components by using a reference mechanism. When a
+ * {@link DescribableComponent} is encountered for the first time, it is fully serialized including its unique
+ * identifier ({@code _ref}). Any subsequent occurrences of the same component instance are replaced with a reference
+ * object containing a {@code $ref} field pointing to the original component's identifier and a {@code _type} field
+ * indicating the component's type. This prevents infinite recursion and {@link StackOverflowError} when describing
+ * components with circular dependencies.
+ * <p>
+ * Example JSON with a circular reference:
+ * <pre>
+ * {
+ *   "component": {
+ *     "_ref": "12345",
+ *     "_type": "MyComponent",
+ *     "name": "First Component",
+ *     "reference": {
+ *       "_ref": "67890",
+ *       "_type": "MyComponent",
+ *       "name": "Second Component",
+ *       "backReference": {
+ *         "$ref": "12345",
+ *         "_type": "MyComponent"
+ *       }
+ *     }
+ *   }
+ * }
+ * </pre>
+ *
+ * @author Mateusz Nowak
+ * @since 5.0.0
+ */
+public class JacksonComponentDescriptor implements ComponentDescriptor {
+
+    private final ObjectMapper objectMapper;
+    private final ObjectNode rootNode;
+    private final Map<DescribableComponent, String> processedComponents;
+
+    /**
+     * Constructs a new {@code JacksonComponentDescriptor} with a default {@link ObjectMapper}.
+     */
+    public JacksonComponentDescriptor() {
+        this(new ObjectMapper());
+    }
+
+    /**
+     * Constructs a new {@code JacksonComponentDescriptor} with the provided {@link ObjectMapper}.
+     *
+     * @param objectMapper The ObjectMapper to use for JSON serialization.
+     */
+    public JacksonComponentDescriptor(ObjectMapper objectMapper) {
+        this(objectMapper.disable(SerializationFeature.FAIL_ON_EMPTY_BEANS), new IdentityHashMap<>());
+    }
+
+    /**
+     * Private constructor used for creating nested descriptors that share the processed components map.
+     *
+     * @param objectMapper        The ObjectMapper to use for JSON serialization.
+     * @param processedComponents Map containing already processed components and their IDs.
+     */
+    private JacksonComponentDescriptor(ObjectMapper objectMapper,
+                                       Map<DescribableComponent, String> processedComponents) {
+        this.objectMapper = objectMapper;
+        this.rootNode = objectMapper.createObjectNode();
+        this.processedComponents = processedComponents;
+    }
+
+    @Override
+    public void describeProperty(@Nonnull String name, @Nonnull Object object) {
```

Comment:
Similar as with the file system version, I don't understand why the collection and map describe methods have a null-check, and this one doesn't.

Reply:
For collection/map is needed, because we iterate over them (cause NullPointerException). For object, it works OK.